crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,618 @@
+    /// then the raw component id will be returned instead.
+    pub fn display(&self, components: &Components) -> String {
+        let component_names: String = self
+            .component_ids()
+            .iter()
+            .map(|id| match components.get_info(*id) {
+                Some(info) => info.name().to_owned(),
+                None => format!("{:?}", id),
+            })
+            .reduce(|acc, s| format!("{}, {}", acc, s))
+            .unwrap_or_default();
+
+        match self {
+            UntypedArchetypeStatement::AllOf(_) => format!("AllOf({component_names})"),
+            UntypedArchetypeStatement::AnyOf(_) => format!("AnyOf({component_names})"),
+            UntypedArchetypeStatement::AtMostOneOf(_) => format!("AtMostOneOf({component_names})"),
+            UntypedArchetypeStatement::NoneOf(_) => format!("NoneOf({component_names})"),
+            UntypedArchetypeStatement::Only(_) => format!("Only({component_names})"),
+        }
+    }
+
+    /// Test if this statement is true for the provided set of [`ComponentId`]s.
+    pub fn test(&self, component_ids: &HashSet<ComponentId>) -> bool {
+        match self {
+            UntypedArchetypeStatement::AllOf(required_ids) => {
+                for required_id in required_ids {
+                    if !component_ids.contains(required_id) {
+                        return false;
+                    }
+                }
+                true
+            }
+            UntypedArchetypeStatement::AnyOf(desired_ids) => {
+                for desired_id in desired_ids {
+                    if component_ids.contains(desired_id) {
+                        return true;
+                    }
+                }
+                false
+            }
+            UntypedArchetypeStatement::AtMostOneOf(exclusive_ids) => {
+                let mut found_previous = false;
+                for exclusive_id in exclusive_ids {
+                    if component_ids.contains(exclusive_id) {
+                        if found_previous {
+                            return false;
+                        }
+                        found_previous = true;
+                    }
+                }
+                true
+            }
+            UntypedArchetypeStatement::NoneOf(forbidden_ids) => {
+                for forbidden_id in forbidden_ids {
+                    if component_ids.contains(forbidden_id) {
+                        return false;
+                    }
+                }
+                true
+            }
+            UntypedArchetypeStatement::Only(only_ids) => {
+                for component_id in component_ids {
+                    if !only_ids.contains(component_id) {
+                        return false;
+                    }
+                }
+                true
+            }
+        }
+    }
+}
+
+/// A list of [`ArchetypeInvariant`]s to be stored on a [`World`].
+#[derive(Default)]
+pub struct ArchetypeInvariants {
```

Comment:
I'm wondering if `ArchetypeInvariants` should be public. After all, is a `pub(crate)` field of `World`, rendering it inaccessible.

Reply:
Unfortunately, the type has to be `pub` since it is returned from the world with `World.archetype_invariants`.

Reply:
(It's also used incidentally in several public utility methods)

Reply:
Could there be any advantage in manually calling `ArchetypeInvariants::test_archetype`? If that's the case, it could return a `Result` instead of panicking, so the user can handle the failure case. The in-engine usage could simply use `expect` and panic at the caller site.